eax/src/online.rs

```diff
@@ -0,0 +1,432 @@
+/// let mut cipher = Eax::<Aes256, Decrypt>::with_key_and_nonce(key, nonce);
+///
+/// cipher.update_assoc(b"tampered");
+/// cipher.decrypt_unauthenticated(&mut cloned);
+/// let res = cipher.finish(&tag);
+///
+/// assert_eq!(res, Err(Error));
+/// ```
+///
+/// [<sup>1</sup>]: https://en.wikipedia.org/wiki/Online_algorithm
+/// [`Eax`]: ../struct.Eax.html
+/// [`Decrypt`]: struct.Decrypt.html
+/// [`finish`]: #method.finish
+pub struct Eax<Cipher, Op>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+    Op: CipherOp,
+{
+    imp: EaxImpl<Cipher>,
+    /// Denotes whether this stream is used for encryption or decryption.
+    marker: PhantomData<Op>,
+}
+
+impl<Cipher, Op> Eax<Cipher, Op>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+    Op: CipherOp,
+{
+    /// Creates a stateful EAX instance that is capable of processing both
+    /// the associated data and the plaintext in an "on-line" fashion.
+    pub fn with_key_and_nonce(key: &Key<Cipher>, nonce: &Nonce<Cipher::BlockSize>) -> Self {
+        let imp = EaxImpl::<Cipher>::with_key_and_nonce(key, nonce);
+
+        Self {
+            imp,
+            marker: PhantomData,
+        }
+    }
+
+    /// Process the associated data (AD).
+    #[inline]
+    pub fn update_assoc(&mut self, aad: &[u8]) {
+        self.imp.update_assoc(aad);
+    }
+
+    /// Derives the tag from the encrypted/decrypted message so far.
+    ///
+    /// If the encryption/decryption operation is finished, [`finish`] method
+    /// *must* be called instead.
+    ///
+    ///[`finish`]: #method.finish
+    #[inline]
+    pub fn tag_clone(&self) -> Tag {
+        self.imp.tag_clone()
+    }
+}
+
+impl<Cipher> Eax<Cipher, Encrypt>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+{
+    /// Applies encryption to the plaintext.
+    #[inline]
+    pub fn encrypt(&mut self, msg: &mut [u8]) {
+        self.imp.encrypt(msg)
+    }
+
+    /// Finishes the encryption stream, returning the derived tag.
+    ///
+    /// This *must* be called after the stream encryption is finished.
+    #[must_use = "tag must be saved to later verify decrypted data"]
+    #[inline]
+    pub fn finish(self) -> Tag {
+        self.imp.tag()
+    }
+}
+
+impl<Cipher> Eax<Cipher, Decrypt>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+{
+    /// Applies decryption to the ciphertext **without** verifying the
+    /// authenticity of decrypted message.
+    ///
+    /// To correctly verify the authenticity, use the [`finish`] associated
+    /// function.
+    ///
+    /// [`finish`]: #method.finish
+    #[inline]
+    pub fn decrypt_unauthenticated(&mut self, msg: &mut [u8]) {
```

Comment:
Would still prefer `hazmat` somewhere in here. IMO it's not unclear this can break IND-CCA(3) security.

Reply:
I did add 3b180b9c0927e599e8d396ff8683f6b1bfeb05c3, which hopefully addresses this. I'm not a cryptographer so I'm not really sure that I understood/phrased the IND-CCA security in this context; let me know if the current phrasing is correct and acceptable.

Reply:
Looks good, thanks!